app/controllers/concerns/two_factor_authenticatable_methods.rb

```diff
@@ -234,6 +242,23 @@ def mark_user_session_authenticated_analytics(authentication_type)
     )
   end
 
+  def mfa_device_type(auth_method:)
+    return 'otp' if auth_method == TwoFactorAuthenticatable::AuthMethod::SMS
+
+    auth_method
+  end
+
+  def format_failures(auth_method:, result:)
+    if auth_method == TwoFactorAuthenticatable::AuthMethod::PIV_CAC
```

Comment:
this is pretty goofy, but for some reason the piv_cac form does not create errors in the same way that other forms do. it seems like it ends up with NO `error_details` attribute, and instead creates an `errors` attribute that always looks like this: `{ type: 'token.invalid' }` i tried to fix this higher up on the code chain (like in the form itself), but since the FormResponse generates error_details, that felt too invasive.

i could add this to the tracker code if we thought this was the only place this would really be invoked? i just wasn't sure. wdyt?

Reply:
i would also be interested in just fixing the way piv cac does errors, if we don't think that will break anything!

Reply:
here is the [ticket](https://gitlab.login.gov/lg-teams/Melba/protocols-backlog/-/issues/162) and the [pr](https://github.com/18F/identity-idp/pull/12052) for fixing this! once it's merged in, i will fix it here

Reply:
fixed in https://github.com/18F/identity-idp/pull/12052/commits/ccd1f9e88c1708c0292f9a1305f58ac749f00323 !!!!!